### PR TITLE
add scripts of entailment (#348)

### PR DESCRIPTION
**Pull request by hhexiy**
## Description ##
@linmx0130 's Implementation of the entailment model from [A Decomposable Attention Model for Natural Language Inference](https://arxiv.org/abs/1606.01933).
Run on SNLI dataset. Dataset is omitted and can be downloaded from https://nlp.stanford.edu/projects/snli/.

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
some hyperparameters are different:
 1. Xavier initializer is utilized instead of Gaussian initializer. The model with later one does not converge.
 2. Dropout layers are removed.
 3. Learning rate is smaller.

### CODE REVIEW

**Comment by szha** [comment]
Thanks @linmx0130 for the work.

File: scripts/natural_language_inference/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# NLIGluon
```
Comment (szha) [inline scripts/natural_language_inference/README.md]: This doc needs to be made into a model zoo page for natural language inference (e.g. natural_language_inference/index.rst). See examples in https://github.com/dmlc/gluon-nlp/tree/master/scripts

File: scripts/natural_language_inference/decomposable_atten.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+class IntraSentenceAtten(gluon.Block):
+    """
+    Intra Sentence Attantion block.
+    """
+    def __init__(self, inp_size: int, hidden_size: int, max_length=10, **kwargs):
+        super(IntraSentenceAtten, self).__init__(**kwargs)
+        self.max_length = max_length
+        self.hidden_size = hidden_size
+        with self.name_scope():
+            self.func = nn.Sequential()
```
Comment (szha) [inline scripts/natural_language_inference/decomposable_atten.py]: Use HybridSequential

Comment (szha) [reply scripts/natural_language_inference/decomposable_atten.py]: Also, try to use more informative names.

File: scripts/natural_language_inference/decomposable_atten.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+# pylint: disable=C0103,E1101,R0914
```
Comment (szha) [inline scripts/natural_language_inference/decomposable_atten.py]: Remove disable and fix the problems if possible. If not, please state the reasons

File: scripts/natural_language_inference/decomposable_atten.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+decomposable_atten.py
+
+Part of NLI script of gluon-nlp. Implementation of Decomposable Attention.
+Copyright 2018 Mengxiao Lin <linmx0130@gmail.com>.
+"""
+
+from mxnet import nd, gluon
+from mxnet.gluon import nn
+
+class IntraSentenceAtten(gluon.Block):
```
Comment (szha) [inline scripts/natural_language_inference/decomposable_atten.py]: It should be feasible to make this a HybridBlock for better performance.

File: scripts/natural_language_inference/decomposable_atten.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            self.func.add(nn.Dense(hidden_size, in_units=hidden_size))
+
+    def forward(self, *args):
+        feature_a = args[0]
+        batch_size, length, inp_size = feature_a.shape
+        tilde_a = self.func(
+            feature_a.reshape(batch_size * length, inp_size)).reshape(
+                batch_size, length, self.hidden_size)
+        e_matrix = nd.linalg.gemm2(A=tilde_a, B=tilde_a.transpose([0, 2, 1]))
+        alpha = nd.linalg.gemm2(nd.softmax(e_matrix), tilde_a)
```
Comment (szha) [inline scripts/natural_language_inference/decomposable_atten.py]: dot?

File: scripts/natural_language_inference/nlidataset.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+nlidataset.py
+
+Part of NLI script in gluon-nlp.
+Copyright 2018 Mengxiao Lin <linmx0130@gmail.com>
+"""
+
+import json
+import mxnet.gluon.data as gdata
+
+class NLIDataItem:
```
Comment (szha) [inline scripts/natural_language_inference/nlidataset.py]: In gluonnlp, we intentionally avoid using an item class so that it's more flexible. You may use a simple tuple to achieve the same and put the transformation logic in transform functions.

File: scripts/natural_language_inference/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+The model is following [A Decomposable Attention Model for Natural Language Inference](https://arxiv.org/abs/1606.01933).
+
+However, some hyperparameters are different:
+1. Xavier initializer is utilized instead of Gaussian initializer. The model with later one does not converge.
+2. Dropout layers are removed.
+3. Learning rate is smaller.
+
+## Usage
+Put SNLI dataset into `data/`, then train the model with
```
Comment (szha) [inline scripts/natural_language_inference/README.md]: How does one obtain SNLI dataset? Does it allow redistribution? If so, we can add it as a public dataset in GluonNLP's data API.

File: scripts/natural_language_inference/utils.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+import gluonnlp as nlp
+import nltk
+
+def tokenize_and_index(sentence: str, embedding: nlp.embedding.TokenEmbedding):
+    """
+    Tokenize sentence with NLTK and obtain token index in provided embedding object.
+    """
+    words = nltk.word_tokenize(sentence.lower())
+    return [embedding.token_to_idx[t] for t in words]
```
Comment (szha) [inline scripts/natural_language_inference/utils.py]: GluonNLP supports nltk integration for tokenization. Also, it's rarely necessary to use TokenEmbedding this way. See http://gluon-nlp.mxnet.io/examples/word_embedding/word_embedding.html#Attaching-word-embeddings

File: scripts/natural_language_inference/main_intra.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    def __init__(self, word_embed_size, hidden_size, **kwargs):
+        super(Network, self).__init__(**kwargs)
+        self.word_embed_size = word_embed_size
+        self.hidden_size = hidden_size
+        with self.name_scope():
+            self.lin_proj = gluon.nn.Dense(hidden_size, in_units=word_embed_size, activation='relu')
+            self.intra_atten = IntraSentenceAtten(hidden_size, hidden_size)
+            self.model = DecomposableAtten(hidden_size*2, hidden_size, 3)
+
+    def forward(self, *args):
```
Comment (bikestra) [inline scripts/natural_language_inference/main_intra.py]: Again, why not
```suggestion
    def forward(self, sentence1, sentence2, *args):
```
?

File: scripts/natural_language_inference/main_intra.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    emb_layer = mx.gluon.nn.Embedding(len(embedding.idx_to_token),
+                                      embedding.idx_to_vec[0].size)
+    emb_layer.initialize()
+    emb_layer.weight.set_data(embedding.idx_to_vec)
+    padder = nlp.data.batchify.Pad()
+    model.save_parameters(
+        os.path.join(args.checkpoints, 'epoch-0.gluonmodel'))
+    for epoch in range(1, args.maximum_iter + 1):
+        access_key = list(range(len(train_set)))
+        np.random.shuffle(access_key)
```
Comment (bikestra) [inline scripts/natural_language_inference/main_intra.py]: I do not see any seeding of random number generators. It would be great to have all RNGs seeded so that results are reproducible.

File: scripts/natural_language_inference/main_intra.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        os.mkdir(args.checkpoints)
+    emb_layer = mx.gluon.nn.Embedding(len(embedding.idx_to_token),
+                                      embedding.idx_to_vec[0].size)
+    emb_layer.initialize()
+    emb_layer.weight.set_data(embedding.idx_to_vec)
+    padder = nlp.data.batchify.Pad()
+    model.save_parameters(
+        os.path.join(args.checkpoints, 'epoch-0.gluonmodel'))
+    for epoch in range(1, args.maximum_iter + 1):
+        access_key = list(range(len(train_set)))
```
Comment (bikestra) [inline scripts/natural_language_inference/main_intra.py]: I think shallow copying `train_set` and shuffling it is easier to read and as efficient than this. If we are staying with this, the variable name `access_key` is a bit unconventional. I would suggest something like `data_indices` so that it is more clear what this list is about.

File: scripts/natural_language_inference/main_intra.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        counter = print_period
+        acc_loss = nd.array([0.], ctx=ctx)
+        acc_acc = nd.array([0.], ctx=ctx)
+        while idx + batch_size <= len(access_key):
+            sen1 = []
+            sen2 = []
+            label = []
+            for i in range(batch_size):
+                item = train_set[access_key[idx + i]]
+                sen1.append(tokenize_and_index(item.sentence1, embedding))
```
Comment (bikestra) [inline scripts/natural_language_inference/main_intra.py]: Instead of doing the data processing inside of the training loop, I would suggest to write a `glounnlp.data.DataStream` object so that training logic is decoupled from data processing logic. This will also help us reduce the duplication of code in `train_network()` and `test_network()`.

File: scripts/natural_language_inference/decomposable_atten.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        super(IntraSentenceAtten, self).__init__(**kwargs)
+        self.max_length = max_length
+        self.hidden_size = hidden_size
+        with self.name_scope():
+            self.func = nn.Sequential()
+            self.func.add(nn.Dense(hidden_size, in_units=inp_size, activation='relu'))
+            self.func.add(nn.Dense(hidden_size, in_units=hidden_size, activation='relu'))
+            self.func.add(nn.Dense(hidden_size, in_units=hidden_size))
+
+    def forward(self, *args):
```
Comment (bikestra) [inline scripts/natural_language_inference/decomposable_atten.py]: Why not 
```suggestion
    def forward(self, feature_a, *args):
```
rather than using args[0]?

File: scripts/natural_language_inference/decomposable_atten.py (hunk 5 above)
Comment (hhexiy) [reply scripts/natural_language_inference/decomposable_atten.py]: we need matrix multiplication here.

**Comment by mli** [comment]
Job PR-404/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-404/17/index.html

File: scripts/natural_language_inference/utils.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+# specific language governing permissions and limitations
+# under the License.
+# Copyright 2018 Mengxiao Lin <linmx0130@gmail.com>.
+
+"""
+Utility functions.
+"""
+
+import logging
+
```
Comment (eric-haibin-lin) [inline scripts/natural_language_inference/utils.py]: This also appeared in scripts/machine_translation/utils.py. It's not urgent but we should probably reuse code by moving this to gluonnlp.utils instead of duplication.

File: scripts/natural_language_inference/decomposable_attention.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+                self.intra_attention = IntraSentenceAttention(hidden_size, hidden_size, dropout)
+                input_size = hidden_size * 2
+            else:
+                self.intra_attention = None
+                input_size = hidden_size
+            self.model = DecomposableAttention(input_size, hidden_size, 3, dropout)
+
+    def hybrid_forward(self, F, sentence1, sentence2):
+        """
+        Model forward definition
```
Comment (eric-haibin-lin) [inline scripts/natural_language_inference/decomposable_attention.py]: Please document input output data shapes

File: scripts/natural_language_inference/decomposable_attention.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            self.intra_attn_emb = nn.HybridSequential()
+            self.intra_attn_emb.add(self.dropout_layer)
+            self.intra_attn_emb.add(nn.Dense(hidden_size, in_units=inp_size,
+                                             activation='relu', flatten=False))
+            self.intra_attn_emb.add(self.dropout_layer)
+            self.intra_attn_emb.add(nn.Dense(hidden_size, in_units=hidden_size,
+                                             activation='relu', flatten=False))
+
+    def hybrid_forward(self, F, feature_a):
+        # batch_size, length, inp_size = feature_a.shape
```
Comment (eric-haibin-lin) [inline scripts/natural_language_inference/decomposable_attention.py]: Same comment - please document input output data shapes
